File: src/main/java/com/github/unchama/seichiassist/menus/minestack/CategorizedMineStackMenu.kt
package com.github.unchama.seichiassist.menus.minestack

import com.github.unchama.itemstackbuilder.SkullItemStackBuilder
import com.github.unchama.menuinventory.*
import com.github.unchama.menuinventory.slot.button.Button
import com.github.unchama.menuinventory.slot.button.action.ClickEventFilter
import com.github.unchama.menuinventory.slot.button.action.FilteredButtonEffect
import com.github.unchama.seichiassist.MineStackObjectList
import com.github.unchama.seichiassist.Schedulers
import com.github.unchama.seichiassist.SkullOwnerReference
import com.github.unchama.seichiassist.SkullOwners
import com.github.unchama.seichiassist.menus.CommonButtons
import com.github.unchama.seichiassist.minestack.MineStackObjectCategory
import com.github.unchama.seichiassist.minestack.category
import com.github.unchama.targetedeffect.TargetedEffect
import com.github.unchama.targetedeffect.computedEffect
import com.github.unchama.targetedeffect.player.FocusedSoundEffect
import com.github.unchama.targetedeffect.sequentialEffect
import com.github.unchama.targetedeffect.unfocusedEffect
import com.github.unchama.util.collection.mapValues
import org.bukkit.ChatColor.*
import org.bukkit.Sound
import org.bukkit.entity.Player
import kotlin.math.ceil

object CategorizedMineStackMenu {
  private const val mineStackObjectPerPage = 9 * 5

  private suspend fun Player.computeMenuLayout(category: MineStackObjectCategory, page: Int): IndexedSlotLayout {
    val categoryItemList = MineStackObjectList.minestacklist!!.filter { it.category() === category }
    val totalNumberOfPages = ceil(categoryItemList.size / 45.0).toInt()

    // オブジェクトリストが更新されるなどの理由でpageが最大値を超えてしまった場合、最後のページを計算する
    if (page >= totalNumberOfPages) return computeMenuLayout(category, totalNumberOfPages - 1)

    // カテゴリ内のMineStackアイテム取り出しボタンを含むセクション
    val categorizedItemSection =
        IndexedSlotLayout(
            categoryItemList.drop(mineStackObjectPerPage * page).take(mineStackObjectPerPage)
                .withIndex()
                .mapValues { with (MineStackButtons) { getMineStackItemButtonOf(it) } }
        )

    // ページ操作等のボタンを含むレイアウトセクション
    val uiOperationSection = run {
      fun buttonToTransferTo(page: Int, skullOwnerReference: SkullOwnerReference) = Button(
          SkullItemStackBuilder(skullOwnerReference)
              .title("$YELLOW$UNDERLINE${BOLD}MineStack${page + 1}ページ目へ")
              .lore(listOf("$RESET$DARK_RED${UNDERLINE}クリックで移動"))
              .build(),
          FilteredButtonEffect(ClickEventFilter.LEFT_CLICK) {
            sequentialEffect(
                FocusedSoundEffect(Sound.BLOCK_FENCE_GATE_OPEN, 1.0f, 0.1f),
                forCategory(category, page).open
            )
          }
      )

      val stickMenuButtonSection = singleSlotLayout { (9 * 5) to CommonButtons.openStickMenu }

      val previousPageButtonSection = if (page > 0) {
        singleSlotLayout { 9 * 5 + 7 to buttonToTransferTo(page - 1, SkullOwners.MHF_ArrowUp) }
      } else emptyLayout

      val nextPageButtonSection = if (page + 1 < totalNumberOfPages) {
        singleSlotLayout { 9 * 5 + 8 to buttonToTransferTo(page + 1, SkullOwners.MHF_ArrowDown) }
      } else emptyLayout

      combinedLayout(
          stickMenuButtonSection,
          previousPageButtonSection,
          nextPageButtonSection
      )
    }

    // 自動スタック機能トグルボタンを含むセクション
    val autoMineStackToggleButtonSection = singleSlotLayout {
      (9 * 5 + 4) to with(MineStackButtons) { computeAutoMineStackToggleButton() }
    }

    return combinedLayout(
        categorizedItemSection,
        uiOperationSection,
        autoMineStackToggleButtonSection
    )
  }

  /**
   * カテゴリ別マインスタックメニューで[page]ページ目の[Menu]
   */
  fun forCategory(category: MineStackObjectCategory, page: Int = 0): Menu = object: Menu {
    override val open: TargetedEffect<Player> = computedEffect { player ->
      val session = MenuInventoryView(
          6.rows(),
          "$DARK_BLUE${BOLD}MineStack - ${category.uiLabel} (${page}ページ目)"
      ).createNewSession()

      sequentialEffect(
          session.openEffectThrough(Schedulers.sync),
          unfocusedEffect { session.overwriteViewWith(player.computeMenuLayout(category, page)) }
      )
    }
  }
}

File: src/main/scala/com/github/unchama/seichiassist/menus/minestack/CategorizedMineStackMenu.scala
package com.github.unchama.seichiassist.menus.minestack

import cats.effect.IO
import com.github.unchama.generic.MapExtra
import com.github.unchama.itemstackbuilder.{SkullItemStackBuilder, SkullOwnerReference}
import com.github.unchama.menuinventory._
import com.github.unchama.menuinventory.router.CanOpen
import com.github.unchama.menuinventory.slot.button.Button
import com.github.unchama.minecraft.actions.OnMinecraftServerThread
import com.github.unchama.seichiassist.SkullOwners
import com.github.unchama.seichiassist.menus.CommonButtons
import com.github.unchama.seichiassist.subsystems.gachaprize.GachaPrizeAPI
import com.github.unchama.seichiassist.subsystems.minestack.MineStackAPI
import com.github.unchama.seichiassist.subsystems.minestack.domain.minestackobject.MineStackObjectCategory
import com.github.unchama.targetedeffect.{DeferredEffect, TargetedEffect}
import org.bukkit.ChatColor._
import org.bukkit.entity.Player
import org.bukkit.inventory.ItemStack

object CategorizedMineStackMenu {

  class Environment(
    implicit val ioCanOpenMineStackMainMenu: IO CanOpen MineStackMainMenu.type,
    val ioCanOpenCategorizedMenu: IO CanOpen CategorizedMineStackMenu,
    val ioCanOpenSelectItemColorMenu: IO CanOpen MineStackSelectItemColorMenu,
    val onMainThread: OnMinecraftServerThread[IO],
    val mineStackAPI: MineStackAPI[IO, Player, ItemStack],
    implicit val gachaPrizeAPI: GachaPrizeAPI[IO, ItemStack, Player]
  )

}

/**
 * カテゴリ別マインスタックメニューで [pageIndex] + 1 ページ目の[Menu]
 */
case class CategorizedMineStackMenu(category: MineStackObjectCategory, pageIndex: Int = 0)
    extends Menu {

  import com.github.unchama.menuinventory.syntax._
  import eu.timepit.refined.auto._

  /**
   * マインスタックオブジェクトボタンを置くセクションの行数
   */
  private val objectSectionRows = 5

  override type Environment = CategorizedMineStackMenu.Environment

  override val frame: MenuFrame = MenuFrame(
    (objectSectionRows + 1).chestRows,
    s"$DARK_BLUE${BOLD}MineStack(${category.uiLabel})"
  )

  private def mineStackMainMenuButtonSection(
    implicit ioCanOpenMineStackMainMenu: IO CanOpen MineStackMainMenu.type
  ): Seq[(Int, Button)] =
    Seq(
      ChestSlotRef(5, 0) -> CommonButtons.transferButton(
        new SkullItemStackBuilder(SkullOwners.MHF_ArrowLeft),
        "MineStackメインメニューへ",
        MineStackMainMenu
      )
    )

  // ページ操作等のボタンを含むレイアウトセクション
  def uiOperationSection(totalNumberOfPages: Int)(category: MineStackObjectCategory, page: Int)(
    implicit environment: Environment
  ): Seq[(Int, Button)] = {
    import environment._

    def buttonToTransferTo(pageIndex: Int, skullOwnerReference: SkullOwnerReference): Button =
      CommonButtons.transferButton(
        new SkullItemStackBuilder(skullOwnerReference),
        s"MineStack${pageIndex + 1}ページ目へ",
        CategorizedMineStackMenu(category, pageIndex)
      )

    val previousPageButtonSection =
      MapExtra.when(page > 0)(
        Map(ChestSlotRef(5, 7) -> buttonToTransferTo(page - 1, SkullOwners.MHF_ArrowUp))
      )

    val nextPageButtonSection =
      MapExtra.when(page + 1 < totalNumberOfPages)(
        Map(ChestSlotRef(5, 8) -> buttonToTransferTo(page + 1, SkullOwners.MHF_ArrowDown))
      )

    mineStackMainMenuButtonSection ++ previousPageButtonSection ++ nextPageButtonSection
  }

  override def open(
    implicit environment: CategorizedMineStackMenu.Environment,
    ctx: LayoutPreparationContext,
    onMainThread: OnMinecraftServerThread[IO]
  ): TargetedEffect[Player] = DeferredEffect {

    for {
      categoryGroupCount <-
        environment
          .mineStackAPI
          .mineStackObjectList
          .getAllObjectGroupsInCategory(category)
          .map(_.length)
    } yield {
      val totalNumberOfPages = Math.ceil(categoryGroupCount / 45.0).toInt

      // オブジェクトリストが更新されるなどの理由でpageが最大値を超えてしまった場合、
      // 最後のページをopenする作用を返す
      if (pageIndex >= totalNumberOfPages)
        CategorizedMineStackMenu(category, totalNumberOfPages - 1).open
      else super.open
    }
  }

  override def computeMenuLayout(
    player: Player
  )(implicit environment: Environment): IO[MenuSlotLayout] = {
    import cats.implicits._
    import environment._

    val mineStackObjectPerPage = objectSectionRows.chestRows.slotCount

    val playerMineStackButtons = MineStackButtons(player)
    import playerMineStackButtons._

    // 自動スタック機能トグルボタンを含むセクションの計算
    val autoMineStackToggleButtonSectionComputation =
      List(ChestSlotRef(5, 4) -> computeAutoMineStackToggleButton).traverse(_.sequence)

    for {
      categoryGroups <- mineStackAPI.mineStackObjectList.getAllObjectGroupsInCategory(category)
      totalNumberOfPages = Math.ceil(categoryGroups.length / 45.0).toInt
      categorizedItemSection <-
        categoryGroups // カテゴリ内のMineStackアイテム取り出しボタンを含むセクションの計算
          .slice(
            mineStackObjectPerPage * pageIndex,
            mineStackObjectPerPage * pageIndex + mineStackObjectPerPage
          )
          .traverse(getMineStackGroupButtonOf(_, pageIndex))
          .map(_.zipWithIndex.map(_.swap))
      autoMineStackToggleButtonSection <- autoMineStackToggleButtonSectionComputation
    } yield {
      val combinedLayout =
        uiOperationSection(totalNumberOfPages)(category, pageIndex)
          .++(categorizedItemSection)
          .++(autoMineStackToggleButtonSection)

      MenuSlotLayout(combinedLayout: _*)
    }
  }
}
